### PR TITLE
Issue #3342: Indentation: Lambda re-implementation

### LINKED ISSUE #3342: Indentation: lambda don't respect lineWrappingIndentation when forceStrictCondition is false
/var/tmp $ javac Test.java

```
```

/var/tmp $ cat Test.java

```
package com.github.aruberto;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public class Test {

    public static void someFunction(SomeUtilClass util) {

        // This fails indentation rule
        util.myLambdaUtil("FIRST_ARG",
                          (string) -> System.out.println(string.trim()), //line 13
                          "SECOND_ARG",
                          () -> "WHAT WHAT!");      //line 15

        // This passes indentation rule
        util.myLambdaUtil("FIRST_ARG",
            (string) -> System.out.println(string.trim()),
                          "SECOND_ARG",
            () -> "WHAT WHAT!");

        // This fails indentation rule
        Function<String, String> someFunction1 =
                (string) -> {                       // line 25
                    if (string.contains("abc")) {
                        return "SWEET!";
                    } else if (string.contains("123")) {
                        return "COOL!";
                    } else {
                        return "BOO!";
                    }
                };                                 // line 33

        // This passes indentation rule
        Function<String, String> someFunction2 =
        (string) -> {
            if (string.contains("abc")) {
                return "SWEET!";
            } else if (string.contains("123")) {
                return "COOL!";
            } else {
                return "BOO!";
            }
        };
    }

    interface SomeUtilClass {

        void myLambdaUtil(String firstArg,
                          Consumer<String> firstLambda,
                          String secondArg,
                          Supplier<String> secondLambda);
    }
}    
```

/var/tmp $ cat config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="charset" value="UTF-8"/>

  <property name="severity" value="warning"/>

  <property name="fileExtensions" value="java, properties, xml"/>

  <module name="FileTabCharacter">
    <property name="eachLine" value="true"/>
  </module>

  <module name="TreeWalker">
    <module name="Indentation">
      <property name="basicOffset" value="4"/>
      <property name="braceAdjustment" value="0"/>
      <property name="caseIndent" value="4"/>
      <property name="throwsIndent" value="8"/>
      <property name="arrayInitIndent" value="4"/>
      <property name="lineWrappingIndentation" value="0"/>
      <property name="forceStrictCondition" value="false"/>
    </module>
  </module>
</module>                                                                         
```

/var/tmp $ java -jar checkstyle-7.0-all.jar -c config.xml Test.java

```
Starting audit...                                                                                                                                                                               
[WARN] Test.java:13: 'lambda arguments' have incorrect indentation level 26, expected level should be one of the following: 8, 12. [Indentation] 
[WARN] Test.java:15: 'lambda arguments' have incorrect indentation level 26, expected level should be one of the following: 8, 12. [Indentation] 
[WARN] Test.java:25: 'lambda arguments' have incorrect indentation level 16, expected level should be 8. [Indentation]                           
[WARN] Test.java:26: 'if' have incorrect indentation level 20, expected level should be 12. [Indentation]                                        
[WARN] Test.java:27: 'if' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                  
[WARN] Test.java:28: 'if rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                                 
[WARN] Test.java:29: 'if' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                  
[WARN] Test.java:30: 'if rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                                 
[WARN] Test.java:31: 'else' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                
[WARN] Test.java:32: 'else rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                               
[WARN] Test.java:33: 'block rcurly' have incorrect indentation level 16, expected level should be 8. [Indentation]                               
Audit done.                                                                                                                                                                                     
```

---

Describe what you expect in details.

Given that forceStrictCondition is false and lineWrappingIndentation = 0, I should be able to indent lambda expressions on wrapping lines at any indent level > 0. 

In First example, the String arg on line 14 is aligned and passes indentation rule (indention of 26) but the Consumer and Supplier args on line 13 and 15 are forced to indent at 8 or 12.

In second example, I define my Function on a new line after assignment operator (lines 25-33). This is indented by Intellij to 16 as it respects current line wrapping settings. But checkstyle want an indentation of 8 for wrapping line and all subsequent lines of lambda expression.
##

### PR DESCRIPTION
**Pull request by Abhishek-kumar09**
Closes #3342,
Covers all the cases described in the discussion thread under #3342,

Diff Regression projects: https://gist.githubusercontent.com/Abhishek-kumar09/717190a7e2e06f4747b9c622692d4b99/raw/e14c56fbb6ca5f77b5bd20da1c9e51f0a99284ee/projects-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Abhishek-kumar09/59f7adaffb4eefc379dd29c23227091b/raw/d156d4bc542f4a342be8fe4cccb1dd0e138a4f61/config.xml

### CODE REVIEW

**Comment by Abhishek-kumar09** [comment]
CI check is failing due to #8725

File: config/checkstyle_input_suppressions.xml [hunk 1]
```diff
@@ -524,6 +524,9 @@
   <suppress checks="Regexp" files="indentation[\\/]indentation[\\/]InputIndentationLambda3.java"/>
   <suppress checks="Regexp" files="indentation[\\/]indentation[\\/]InputIndentationLambda4.java"/>
   <suppress checks="Regexp" files="indentation[\\/]indentation[\\/]InputIndentationLambda5.java"/>
+  <suppress checks="Regexp" files="indentation[\\/]indentation[\\/]InputIndentationLambda6.java"/>
```
Comment (romani) [inline config/checkstyle_input_suppressions.xml:527]: please remove this suppressions and please place comments with config in Input files.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 2]
```diff
@@ -27,6 +27,12 @@
  *
  */
 public class LambdaHandler extends AbstractExpressionHandler {
+    /**
+     * Checks whether the lambda is correctly indented, this variable get its value from checking
+     * the lambda handler's indentation, and it is being used in aligning the lambda's children.
+     * A true value depicts lambda is correctly aligned without giving any errors.
+     */
+    private boolean isIdealIndent;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:35]: `isAcceptableIndent`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 3]
```diff
@@ -36,14 +42,22 @@
      * @param ast the abstract syntax tree
      * @param parent the parent handler
      */
+
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:45]: revert this change

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java [hunk 4]
```diff
@@ -115,7 +115,7 @@ public Valid(InputIndentationCtorCall obj, float arg) { //indent:4 exp:4
     public Valid(InputIndentationCtorCall obj, double arg) { //indent:4 exp:4
       obj. //indent:6 exp:6
           super( //indent:10 exp:10
-            x -> arg); //indent:12 exp:12
+          x -> arg); //indent:10 exp:12,14 warn
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java:118]: please keep existing cases as is, and create new one is required

**Review by romani — CHANGES_REQUESTED** [review]
please share regression report in new way, see details at https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation

items:

File: config/checkstyle_input_suppressions.xml (hunk 1 above)
Comment (Abhishek-kumar09) [reply config/checkstyle_input_suppressions.xml:533]: Yes I had already placed config with the required format in all the Input files. 
Actually the comment rule (in indentation every line should end with comment of exp and actual) and the new rule for all input files(every file should end with `//ok` or `//violation`) is conflicting with each other. That's why I have to push it under suppression file. I think it need to be fixed.
See: https://github.com/checkstyle/checkstyle/issues/8725

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 2 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: `isAcceptableIndent` is already the name of a method. Thus renaming it is creating pmd violation that no same name for method and variable is allowed.
Though I have changed it something more accurate about what it actually does.
This variable checks if the lamda expression is indented properly, and thus upon `yes/no` it states its children expected indent.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 3 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java (hunk 4 above)
Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationCtorCall.java]: Done

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by strkkk** [comment]
@Abhishek-kumar09 unfortunately it is working like this right now. Only contributor's repo is checked out and used to generate report.

I think, oipeline can be enhanced with checking both main and contributor repo and moving patch branch to main repo locally for generation.

File: config/checkstyle_input_suppressions.xml (hunk 1 above)
Comment (romani) [reply config/checkstyle_input_suppressions.xml:533]: ok, got it, very confusing - our bad.

I reopened https://github.com/checkstyle/checkstyle/issues/8725
please do me a favor and send separate PR to fix it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 5]
```diff
@@ -59,6 +73,7 @@ protected IndentLevel getIndentImpl() {
 
         DetailAST parent = getMainAst().getParent();
         if (getParent() instanceof NewHandler) {
+            isLambdaCorrectlyIndented = false;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:76]: This new field isn't final after it is first assigned in the constructor. That means its possible the value can flip after the constructor, and it looks like it can be flipped in multiple areas.

Is it possible usage of this field in one method can go down the `true` path and then another method could use it as the `false` path, and then re-calling the old method will then go down the `false` path the 2nd time through?
This is possible it would lead to some randomness issues in the check throughout a run.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 6]
```diff
@@ -83,7 +98,8 @@ public void checkIndentation() {
         if (parent.getType() != TokenTypes.SWITCH_RULE
                 && getLineStart(firstChild) == expandedTabsColumnNo(firstChild)) {
             final IndentLevel level = getIndent();
-            if (!level.isAcceptable(expandedTabsColumnNo(firstChild))) {
+            if (isAcceptableIndent(firstChild, level)) {
+                isLambdaCorrectlyIndented = false;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:102]: Field name also doesn't make sense to me.

You set lambda to not be correctly idented when the indent is acceptable... I see other instances where it is set to false just because of the parent and nothing to do with the indentation.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 6 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:102]: @rnveach  Yes, there is really a blunder in the naming... It is flipping actual logic...
The method name should be `ifNonAcceptableIndent`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 5 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:76]: @rnveach ,
The value of this variable is used only **once** for suggesting the children's expected indent at line **55**
This method is generally executed after running all the methods regarding checking the Lambda's indent.
The main purpose of this variable is checking if there are indentation error logs for Lamda expression.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 6 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:100]: changes made.

**Comment by rnveach** [comment]
CI is failing and I see now this line is removed: https://github.com/checkstyle/checkstyle/commit/90fa8b8b6f4a8fcea8e63f17de5663852e66c67f#diff-9f344b748f502d515ab6489df6337aa3R76
Is the removal intentional?

**Comment by Abhishek-kumar09** [comment]
> Is the removal intentional?

Yes, this line is removed intentionally. I have created diff reports for spring-integration project and there I suspected wrong errors that should have been modified due to this new change, but it was as it is showing earlier indentation log. 

Actually, I inserted that line somewhere when I was in middle of the development process of this PR due to some reasons, but now this is creating unintentional behaviour.

Thank you for pointing it out., Now the boolean variable is more transparent and understandable too. :)

**Comment by rnveach** [comment]
Custom Regression: http://rveach.no-ip.org/checkstyle/regression/reports/293/
Custom Branch: Your changes + https://github.com/rnveach/checkstyle/commit/884bc35ef31738dd79a5f2a915930c5cd5544085

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 5 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: I looked into my concerns with a custom branch and regression. See https://github.com/checkstyle/checkstyle/pull/8719#issuecomment-692380080 .

No exceptions were thrown that would mean we assigning the field 2 different values. So that alleviates my main concern.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java [hunk 7]
```diff
@@ -92,10 +106,17 @@ public void checkIndentation() {
...
                 logError(getMainAst(), "", expandedTabsColumnNo(getMainAst()), level);
             }
         }
     }
 
+    private boolean isNonAcceptableIndent(DetailAST firstChild, IndentLevel level) {
+        return expandedTabsColumnNo(firstChild) < level.getFirstIndentLevel()
+            || getIndentCheck().isForceStrictCondition()
+               && !level.isAcceptable(expandedTabsColumnNo(firstChild));
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:119]: Minor thing, since `expandedTabsColumnNo(firstChild)` is used twice, you could save the value to a variable.

**Review by rnveach — APPROVED** [review]

**Comment by Abhishek-kumar09** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java (hunk 7 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:119]: I have tried to the same but got unusual error: https://github.com/Abhishek-kumar09/checkstyle/pull/16/files

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:119]: I took a look at the previous changes. The exception isn't fully clear what it was, but I am going to assume you got an NPE because you were trying to use `firstChild` in areas where it was most likely `null`.
That can't be the case here, since you use it for both sides of the `||` condition.

Btw, now that I am looking at this a second time, should the first of the `||` have the following `!getIndentCheck().isForceStrictCondition()` ? Force strict should require that exact tab placement, and it can't be anything besides it.

Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java:119]: I think this logic is fine.
The Indentation is **NonAcceptable** if:
either the child is less than the minimum acceptable indent.
OR
strictCondition is true  and the child is not in the acceptable indent.

> should the first of the || have the following !getIndentCheck().isForceStrictCondition()

If the indent is less than minimum of acceptable indent, then doesnt matter if strict condition is true or false, It should  not be accepted...That's why I think this logic is not required.


> Force strict should require that exact tab placement, and it can't be anything besides it.

Currently It is following all the indent which is a part of accetable indent and nothing more or less that, only discrete acceptable values. I think it is still following the strictlyTrue condition.

Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/LambdaHandler.java]: @rnveach 
Diff report with force strict condition true: https://abhishek-kumar09.github.io/github-actions-report-fetch/
Is it good, We can see here that there are more than 1 acceptable indents in some rare cases with true strict condition....
If it defeats the purpose of `trueForceStrictCondition` then I will change it.

**Comment by romani** [comment]
code is rebased to resolve issue in Travis that is already fixed in master.

**Review by romani — APPROVED** [review]
ok to merge if CI pass

**Comment by Abhishek-kumar09** [comment]
@romani Got execution time-out:
[Step 2/2] The build Checkstyle::IDEA Inspections Pull Request #17076 {buildId=3123224} has been running for more than 50 minutes. Terminating...
image
